### PR TITLE
ROX-13582: Intercepting API calls

### PR DESCRIPTION
**Pull request by 0x656b694d**
## Description

Another piece of #3819.
API call interceptor, that issues various events from configured filters.

`ROX_TELEMETRY_API_WHITELIST` environment variable may be set on the central deployment with a comma separated list of API paths (or "*" for any path except a hardcoded list of ignored ones), which will trigger generic "API Call" events, with three properties:
* Path — the API Path;
* Code — the service response status code;
* User-Agent — the client's User Agent;
* HTTP Method, or something alike computed from gRPC method.

Other implemented events are:
* "Cluster Registered", for when a cluster is posted to the DB;
* "Cluster Initialized", for when a cluster changes its state from `UNINITIALIZED` to something;
* "roxctl", for when an API is called by `roxctl`.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] ~Evaluated and added CHANGELOG entry if required~
- [ ] ~Determined and documented upgrade steps~
- [ ] ~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

## Testing Performed

Unit tests.

Example of an "API Call" event received by Segment:
```json
{
  "context": {
    "Client ID": "2a3e1829-8f84-40c1-a761-006f07a59666",
    "Client Name": "Central",
    "library": {
      "name": "analytics-go",
      "version": "3.0.0"
    }
  },
  "event": "API Call",
  "integrations": {},
  "messageId": "59a4f17d-616f-4588-bfc0-c88890330441",
  "originalTimestamp": "2022-12-13T17:51:42.079249978Z",
  "properties": {
    "Code": 500,
    "Method": "GET",
    "Path": "/api/extensions/scannerdefinitions",
    "User-Agent": "Go-http-client/1.1"
  },
  "receivedAt": "2022-12-13T17:52:18.204Z",
  "sentAt": "2022-12-13T17:52:18.113Z",
  "timestamp": "2022-12-13T17:51:42.170Z",
  "type": "track",
  "userId": "fxVLkxH4kOj9UdfZOP3Rcz8Yw/64sAyZgoMXJ6cGynE=",
  "writeKey": "..."
}
```

### CODE REVIEW

**Comment by 0x656b694d** [comment]
Current dependencies on/for this PR:
* master
  * **PR #4070** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/4070" target="_blank">Graphite</a> 
    * **PR #4071** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/4071" target="_blank">Graphite</a> 
      * **PR #4072** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/4072" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/4072?utm_source=stack-comment).

File: central/telemetry/centralclient/instance_config.go
Comment (dhaus67) [inline central/telemetry/centralclient/instance_config.go:23]: I'd have the same question as in the previous PR within the series: whether we should move this to a more generic, importable location to re-use it within the fleetshard-synchronizer code.

File: central/telemetry/centralclient/interceptors.go
Comment (dhaus67) [inline central/telemetry/centralclient/interceptors.go:49]: (same for `putCluster`): I feel like both of those would be good to be covered by unit tests.

File: pkg/telemetry/phonehome/request_params.go
Comment (dhaus67) [inline pkg/telemetry/phonehome/request_params.go:21]: I think this is more of an edge case and not something we necessarily need to address now:

We provide the option to also use gRPC when the server is exposed behind load balancers which do not necessarily support it (there's https://github.com/stackrox/go-grpc-http1 which covers this).

This means, internally while the API call _may_ be a gRPC one, the protocol it actually uses _may_ be something different (e.g. wss).

Internally, we use this for all sorts of clients (e.g. roxctl, sensor <-> central, etc.). I don't know the full history about this, but probably we ran into some issues within some environments, which lead to us coming up with this solution.

I'm not sure whether there's the desire to also collect the _specific_ protocol used, to further determine the sort of environments customers are using, but it might be worth to consider in the future (essentially, what I'm trying to say is that protocol `gRPC` a.k.a. `http2` _may not be correct_ for all gRPC calls).

File: central/telemetry/centralclient/instance_config.go
Comment (0x656b694d) [reply central/telemetry/centralclient/instance_config.go:23]: This one is really central specific. We've discussed this annotation with @rukletsov, and looks like we want to move it to an env var, as _we_ don't use annotations for configuration today.

File: central/telemetry/centralclient/instance_config.go
Comment (stehessel) [inline central/telemetry/centralclient/instance_config.go:56]: As discussed this should be converted to an env variable.

File: pkg/telemetry/phonehome/request_params.go
Comment (0x656b694d) [reply pkg/telemetry/phonehome/request_params.go:21]: I agree that the existing protocol detection is misleading. And, as was not explicitly requested, I'd rather remove this field from the report.

File: central/telemetry/centralclient/interceptors.go
Comment (stehessel) [inline central/telemetry/centralclient/interceptors.go:29]: This should be based on the API whitelist env variable now, right?

File: central/telemetry/centralclient/interceptors_test.go
Comment (stehessel) [inline central/telemetry/centralclient/interceptors_test.go:12]: I find these tests a little bit hard to follow with the number of asserts. Could you maybe add some comments to make the structure and purpose of the test more obvious?

File: central/telemetry/centralclient/interceptors.go
Comment (stehessel) [inline central/telemetry/centralclient/interceptors.go:20]: I assume the plan is to start with a generic API capture and then replace them more customised events down the road?

**Comment by stehessel** [comment]
LGTM overall, just the minor comments.

File: central/telemetry/centralclient/interceptors.go
Comment (0x656b694d) [reply central/telemetry/centralclient/interceptors.go:20]: We actually start with 3 other events (sensor cluster registration/initialization and roxctl), while the "API Call" is there as a generic configurable "backdoor" for tracking specific calls without recompiling.

**Review by stehessel — APPROVED** [review]

File: pkg/telemetry/phonehome/client_config.go
Comment (rukletsov) [inline pkg/telemetry/phonehome/client_config.go:28]: Here there is no chain and no workflow logic, so I'd move this comment to:
* `track()` describing the workflow logic
* `AddInterceptorFunc()` saying that the order in which interceptors are added might matter

Here I'd describe when an `Interceptor` instance is supposed to return `true` and when `false`.

File: pkg/telemetry/phonehome/client_config.go
Comment (rukletsov) [inline pkg/telemetry/phonehome/client_config.go:134]: We need this for the second time — maybe it's time to extract this into `pkg/httputil/`?

File: pkg/telemetry/phonehome/interceptor.go
Comment (rukletsov) [inline pkg/telemetry/phonehome/interceptor.go:17]: I find it hard to track vars defined in one file and used in other files. IIUC, this is only used to guard write access to `Config.interceptors`. If so, then I think:
* it shall at least be declared in `client_config.go`
* but actually it should be part of `Config`: `Config.interceptorsLock` because otherwise a single mutex is used for all `Config`s (I know we have a singleton but this is not clear here and hence breaks the abstraction)
* `track()` should likely copy `cfg.interceptors` under the same mutex
* probably should be a `sync.RWMutex`

File: pkg/telemetry/phonehome/client_config.go
Comment (rukletsov) [inline pkg/telemetry/phonehome/client_config.go:31]: Looks like `props` is only expected to return values,, any reason it's passed as a function param?

File: pkg/telemetry/phonehome/interceptor.go
Comment (rukletsov) [inline pkg/telemetry/phonehome/interceptor.go:23]: Coming back to the `props` discussion: if you only want a single interceptor to do meaningful work, then `props` should be a retvalue. If you envision interceptor to leave a trace in the same global to their scope `props` object _regardless whether they return true or false_, then it's not the interceptor who should decide on how collisions should be handled.

File: central/telemetry/centralclient/interceptors.go
Comment (rukletsov) [inline central/telemetry/centralclient/interceptors.go:13]: I find it hard to reason about the lifetime, ownership, and initialization of `trackedPaths`. It is initialized in `getInstanceConfig()` which is called once as part of `phonehome.Config` singleton initialization. So there is an implicit assumption that `getInstanceConfig()` can (or should) only be called once. 

I'd suggest to make it part of `Config` and pass into interceptors from the `Config` singleton.

File: central/telemetry/centralclient/interceptors.go
Comment (rukletsov) [inline central/telemetry/centralclient/interceptors.go:17]: If this is only to guard access to `uninitializedClusters`, let's call it so.

File: central/telemetry/centralclient/interceptors.go
Comment (rukletsov) [inline central/telemetry/centralclient/interceptors.go:28]: As someone not familiar with the codebase, I don't understand how `ROX_TELEMETRY_API_WHITELIST` is related to this function. One should bring context from different place to understand that `ROX_TELEMETRY_API_WHITELIST` impacts `apiWhiteList` which in turn impacts `trackedPaths` which is used here (but initialized in a different file).

**Review by rukletsov — CHANGES_REQUESTED** [review]

File: central/telemetry/centralclient/instance_config.go
Comment (0x656b694d) [reply central/telemetry/centralclient/instance_config.go:23]: This is a central specific setting, not to be shared with other telemetry clients.

File: central/main.go
Comment (0x656b694d) [reply central/main.go:545]: It could, but here we've got `basicAuthProvider` available.

File: pkg/telemetry/phonehome/request_params.go
Comment (0x656b694d) [reply pkg/telemetry/phonehome/request_params.go:29]: `grpc.Method()` function have this comment:
```
// Method returns the method string for the server context.  The returned
// string is in the format of "/service/method".
```
I considered consistency with this method.

Go `http.Request` mentions this:
```
// Method specifies the HTTP method (GET, POST, PUT, etc.).
// For client requests, an empty string means GET.
```
That's why `GET` is returned if it's empty. But normally it shouldn't happen.

File: pkg/telemetry/phonehome/client_config.go
Comment (0x656b694d) [reply pkg/telemetry/phonehome/client_config.go:132]: These are methods of the Config struct, so they're here close to the definition.

File: pkg/telemetry/phonehome/client_config.go
Comment (0x656b694d) [reply pkg/telemetry/phonehome/client_config.go:108]: There's no c-tor for Config because there would be too many arguments to pass to it. Instead, a client just fills the required fields of the struct. `interceptors` map is not required to be initialized unless it's going to be used.

File: pkg/telemetry/phonehome/client_config.go
Comment (0x656b694d) [reply pkg/telemetry/phonehome/client_config.go:31]: The idea was that a chain of interceptors (per event type) could augment this map one after another. For example, there could be a filtering interceptor that would filter the requests and possibly add some generic properties, and the second one would add more specific properties.
Originally, I was thinking about adding the current `roxctl` event as the second interceptor for `API Call` event, but then made it a separate event.
Today, there's no need for this complexity, as all the events have a unique interceptor.

File: pkg/telemetry/phonehome/interceptor.go
Comment (0x656b694d) [reply pkg/telemetry/phonehome/interceptor.go:23]: Filtering interceptors in the chain will decide if the event will fire or not.

File: central/telemetry/centralclient/interceptors.go
Comment (0x656b694d) [reply central/telemetry/centralclient/interceptors.go:13]: `trackedPath` is a central specific feature, so it shouldn't be part of `phonehome.Config`. 
`getInstanceConfig()` is a central telemetry configuration initialization function, which initializes this central specific package level variable.

**Pull request merged**
